File: 92.反转链表2/reverseBetween.go
package problem0092

type ListNode struct {
	Val  int
	Next *ListNode
}

// 反转从位置m到n的链表，用一趟扫描完成反转
// 小人愚钝，补上脑瘫做法
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	list := []int{-1}
	t := head
	for t != nil {
		list = append(list, t.Val)
		t = t.Next
	}
	l := head
	for index := 1; l != nil; index++ {
		if m <= index && index <= n {
			l.Val = list[n+m-index]
		}
		l = l.Next
	}
	return head
}

func reverseBetween1(head *ListNode, m int, n int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	// 找到反转头节点的前驱节点
	prev := dummy
	for i := 1; i < m; i++ {
		prev = prev.Next
	}
	// 反转m~n
	cur := prev.Next
	for i := m; i < n; i++ {
		// 核心
		next := cur.Next
		cur.Next = next.Next
		next.Next = prev.Next
		prev.Next = next
	}
	return dummy.Next
}

File: 92.反转链表2/reverseBetween.js
// 反转链表2
// 给你单链表的头指针 head 和两个整数 left 和 right ，其中 left <= right 。请你反转从位置 left 到位置 right 的链表节点，返回 反转后的链表 。
// 链表中节点数目为 n
// 1 <= n <= 500
// 1 <= left <= right <= n

// 由于可能操作头节点，因此需要一个dummy节点
function reverseBetween(head, left, right) {
	const dummy = { next: head }
	// 找到left位置的前一个节点
	let prev = dummy
	for (let i = 1; i < left; i++) {
		prev = prev.next
	}
	// left位置
	let cur = prev.next
	// 备份
	const prevCache = prev
	const curCache = cur
	// 反转left，right位置的节点
	for (let i = left; i <= right; i++) {
		const next = cur.next
		cur.next = prev
		prev = cur
		cur = next
	}
	// 反转完成后，prev成了头节点, cur则是right后面的节点
	prevCache.next = prev
	curCache.next = cur
	return dummy.next
}